### PR TITLE
Replacing SessionAuthentication with ApiKeyAuthentication

### PR DESCRIPTION
**Pull request by AuHau**
Hey there,

so as I said earlier I worked on the replacement of the SessionAuthentication with ApiKeyAuthentication and I am more or less finished! At least all the functionality what I could think of is implemented, I still want to comment bit more the code I wrote, but you could already have a look on it, test it and if you have some objections I could start working on them. 

So what I did?
-   (Backend) Implemented Authentication API - Sign-up/Login/User edit/Password recovery - I changed the HTTP error number of unauthorized requests from 401 to 403, because Tastypie was returning www-authentication header with it and browser was displaying because of it the "HTTP Basic Login window"... In future Tastypie will be also returning 403, but for now I had to fix it.
-   (Frontend) Modal window - Using AngularModalService I created modal window for all the forms I needed.
-   (Frontend) AuthService - I created AuthService for the login, signup etc. Login has "Remember me" feature, which will store the ApiKey into LocalStorage instead of just SessionStorage.
-   Password Recovery - I bit hacked the Django Authentication package and used its PasswordRecoveryForm for generating the token and sending out the email, but I think it is quiet legitimit (and easiest) way of doing it. 

Well I think this is it for now, all the functionality which is currently on master is also in this PR, so that should be OK. I think I managed to quiet follow the design style you started, but if you have some different ideas about how the modal window should look, let me know!

Please check me code (this time it will be quiet more to check than last time :stuck_out_tongue: ) and let me know what you think about it!

Looking forward for your comments! :-)
AuHau

### CODE REVIEW

**Comment by AuHau** [comment]
Just to mention it, this PR is concerning issue #6.

Moreover I would like to suggest, to implement HTTPS for the communication with server. Today it is quiet cheap and easy to implement HTTPS for example with Let's encrypt. I don't have personal experiences with it, but I think that could be easy and good solution for it!

**Comment by AuHau** [comment]
Okey! So now it should be finished :smiley:  I added few things and commented the code.

One thing when you will be deploying it. After you will run migrations to generate ApiKey tables, the current users won't have generated ApiKeys which will resulted in not being able to log in. To generate the ApiKeys use this command `python manage.py backfill_api_keys`, that will generate ApiKeys for all users which doesn't have them.

File: frontend/css/app.scss [hunk 1]
```diff
@@ -557,6 +557,126 @@
...
+            color: white;
+        }
+
+        &.success{
+            background-color: $colorSuccess;
+            color: white;
+        }
+    }
+}
+#fade {
```
Comment (nicbou) [inline frontend/css/app.scss:740]: Shouldn't this be a class?

Comment (AuHau) [reply frontend/css/app.scss:740]: Can be :-) But I assume that there will be only one modal window at the time, so it should not matter that much... Do you want me to still change it?

**Comment by nicbou** [comment]
Hi Adam!

I'm currently looking at your work, both in the backend and in the frontend, and I think you did a marvellous job. I will make a few changes to the CSS to keep the looks consistent, but I think everything is top notch otherwise. I'm glad to finally see some unit tests in there, something I usually only do on paid work :P

I'm going to have a closer look at the code and do some more QA work, as it's a huge change. However, I should be able to merge this fairly soon.

**Comment by AuHau** [comment]
Hey there!

Awesome, that you like it! :-) Sure, sure, let me know if you will need some clarifications, otherwise I am looking forward for your input! 
Btw. I don't think that I wrote any unit tests :laughing: , not sure where do you seem them...

**Pull request merged**